compiler/rustc_data_structures/src/sync.rs

```diff
@@ -360,6 +413,129 @@ cfg_if! {
     }
 }
 
+pub fn for_each_in<T: IntoIterator>(t: T, mut for_each: impl FnMut(T::Item)) {
+    // We catch panics here ensuring that all the loop iterations execute.
+    // This makes behavior consistent with the parallel compiler.
+    let mut panic = None;
+    t.into_iter().for_each(|i| {
+        if let Err(p) = catch_unwind(AssertUnwindSafe(|| for_each(i))) {
+            if panic.is_none() {
+                panic = Some(p);
+            }
+        }
+    });
+    if let Some(panic) = panic {
+        resume_unwind(panic);
+    }
+}
+
+#[macro_export]
+macro_rules! maybe_parallel {
+    ([$tcx: expr], [$tcx_non: expr], $($blocks:expr),*) => {
+        if is_parallel() {
+            parallel!([$tcx], $($blocks),*)
+        } else {
+            // We catch panics here ensuring that all the blocks execute.
+            // This makes behavior consistent with the parallel compiler.
+            let mut panic = None;
+            let tcx = $tcx_non;
+            $(
+                if let Err(p) = ::std::panic::catch_unwind(
+                    ::std::panic::AssertUnwindSafe(|| $blocks(tcx))
+                ) {
+                    if panic.is_none() {
+                        panic = Some(p);
+                    }
+                }
+            )*
+            if let Some(panic) = panic {
+                ::std::panic::resume_unwind(panic);
+            }
```

Comment:
Can this call `for_each_in`?

Reply:
I use a `Vec` to collect all panics and call `resume_unwind` in a loop afterwards